c-dependencies/js-compute-runtime/builtins/object-store.cpp

```diff
@@ -0,0 +1,464 @@
+    }
+  } else {
+    mozilla::Maybe<JS::AutoCheckCannotGC> maybeNoGC;
+    JS::UniqueChars text;
+    char *buf;
+    size_t length;
+
+    if (body_obj && JS_IsArrayBufferViewObject(body_obj)) {
+      // Short typed arrays have inline data which can move on GC, so assert
+      // that no GC happens. (Which it doesn't, because we're not allocating
+      // before `buf` goes out of scope.)
+      maybeNoGC.emplace(cx);
+      JS::AutoCheckCannotGC &noGC = maybeNoGC.ref();
+      bool is_shared;
+      length = JS_GetArrayBufferViewByteLength(body_obj);
+      buf = (char *)JS_GetArrayBufferViewData(body_obj, &is_shared, noGC);
+    } else if (body_obj && JS::IsArrayBufferObject(body_obj)) {
+      bool is_shared;
+      JS::GetArrayBufferLengthAndData(body_obj, &length, &is_shared, (uint8_t **)&buf);
+    } else if (body_obj && URLSearchParams::is_instance(body_obj)) {
+      jsurl::SpecSlice slice = URLSearchParams::serialize(cx, body_obj);
+      buf = (char *)slice.data;
+      length = slice.len;
+    } else {
+      // Convert into a String following https://tc39.es/ecma262/#sec-tostring
+      text = encode(cx, body_val, &length);
+      // 30MB in bytes is the max size allowed for ObjectStore.
+      if (length > 30 * 1024 * 1024) {
+        JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_OBJECT_STORE_PUT_OVER_30_MB);
+        return ReturnPromiseRejectedWithPendingError(cx, args);
+      }
+      if (!text) {
+        return ReturnPromiseRejectedWithPendingError(cx, args);
+      }
+      buf = text.get();
+    }
+
+    BodyHandle body_handle = {INVALID_HANDLE};
+    if (!HANDLE_RESULT(cx, xqd_body_new(&body_handle))) {
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+    }
+
+    if (body_handle.handle == INVALID_HANDLE) {
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+    }
+
+    int result = write_to_body_all(body_handle, buf, length);
+
+    // Ensure that the NoGC is reset, so throwing an error in HANDLE_RESULT
+    // succeeds.
+    if (maybeNoGC.isSome()) {
+      maybeNoGC.reset();
+    }
+
+    if (!HANDLE_RESULT(cx, result)) {
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+    }
+
+    int status =
+        fastly_object_store_insert(object_store_handle(self), key_chars, key_len, body_handle);
+    // Ensure that we throw an exception for all unexpected host errors.
+    if (!HANDLE_RESULT(cx, status)) {
+      return RejectPromiseWithPendingError(cx, result_promise);
+    }
+
+    // The insert was successful so we return a Promise which resolves to undefined
+    JS::RootedValue rval(cx);
+    rval.setUndefined();
+    JS::ResolvePromise(cx, result_promise, rval);
+    args.rval().setObject(*result_promise);
+
+    return true;
+  }
+
+  return false;
+}
+
+const JSFunctionSpec methods[] = {JS_FN("lookup", lookup, 1, JSPROP_ENUMERATE),
+                                  JS_FN("put", put, 1, JSPROP_ENUMERATE), JS_FS_END};
```

Comment:
Why not `get` and `put`?

Reply:
👍 I like your suggestion as it already exists as the counterpart for `put` in JavaScript `Map` objects - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map